webapp/src/main/java/com/box/l10n/mojito/service/smartling/ThirdPartyTextUnitSearchService.java

```diff
@@ -0,0 +1,130 @@
+
+@Component
+public class ThirdPartyTextUnitSearchService {
+
+    private final ImporterCacheService importerCacheService;
+
+    private final ThirdPartyTextUnitRepository thirdPartyTextUnitRepository;
+
+    @Autowired
+    public ThirdPartyTextUnitSearchService(ImporterCacheService importerCacheService,
+                                           ThirdPartyTextUnitRepository thirdPartyTextUnitRepository) {
+        this.importerCacheService = importerCacheService;
+        this.thirdPartyTextUnitRepository = thirdPartyTextUnitRepository;
+    }
+
+    String pluralSuffix = "_other";
+
+    /**
+     * logger
+     */
+    static Logger logger = LoggerFactory.getLogger(ThirdPartyTextUnitSearchService.class);
+
+    List<ThirdPartyTextUnitForBatchImport> convertDTOToBatchImport(Set<ThirdPartyTextUnitDTO> thirdPartyTextUnitDTOSet, boolean isPluralFile) {
+
+        logger.debug("Create caches to map convert to ThirdPartyTextUnitForBatchImport list");
+        LoadingCache<String, Repository> repositoriesCache = importerCacheService.createRepositoriesCache();
+        LoadingCache<Map.Entry<String, Long>, Asset> assetsCache = importerCacheService.createAssetsCache();
+        LoadingCache<Map.Entry<String, Long>, TMTextUnitCurrentVariant> tmTextUnitCurrentVariantCache = importerCacheService.createTmTextUnitCurrentVariantCache();
+
+        logger.debug("Convert to ThirdPartyTextUnitForBatchImport");
+        return thirdPartyTextUnitDTOSet.stream()
+                .filter(logIfFalse(t -> t.getThirdPartyTextUnitId() != null, logger, "Missing mandatory third party text unit id, skip: {}", ThirdPartyTextUnitDTO::getThirdPartyTextUnitId))
+                .filter(logIfFalse(t -> t.getMappingKey() != null, logger, "Missing mandatory mapping key, skip: {}", ThirdPartyTextUnitDTO::getThirdPartyTextUnitId))
+
+                .map(t -> {
+                    ThirdPartyTextUnitForBatchImport thirdPartyTextUnitForBatchImport = new ThirdPartyTextUnitForBatchImport();
+                    thirdPartyTextUnitForBatchImport.setThirdPartyTextUnitId(t.getThirdPartyTextUnitId());
+                    thirdPartyTextUnitForBatchImport.setMappingKey(t.getMappingKey());
+
+                    thirdPartyTextUnitForBatchImport.setRepository(repositoriesCache.getUnchecked(t.getRepositoryName()));
+                    if (thirdPartyTextUnitForBatchImport.getRepository() != null) {
+                        thirdPartyTextUnitForBatchImport.setAsset(assetsCache.getUnchecked(new AbstractMap.SimpleEntry<>(
+                                t.getAssetPath(), thirdPartyTextUnitForBatchImport.getRepository().getId())));
+
+                        if (thirdPartyTextUnitForBatchImport.getAsset() != null) {
+                            TMTextUnitCurrentVariant foundCurrentVariant;
+                            if (isPluralFile) {
+                                foundCurrentVariant = tmTextUnitCurrentVariantCache.getUnchecked(
```

Comment:
only build the key in the if statement (`t.getTmTextUnitName() ` or `t.getTmTextUnitName() + pluralSuffix`) and share the common code

Reply:
This isn't quite clear, is your suggestion to add a null check for `t.getTmTextUnitName()`? Also it is unclear to me what "share the common code" means? The mappingKey is text unit name + #@# + asset path. We've already checked for asset path above and returned the asset so we have asset id and we need to do the check on the text unit name.

Reply:
I'm going to add a filter for `t.getTmTextUnitName()`

Reply:
```
if (isPluralFile) {
                                foundCurrentVariant = tmTextUnitCurrentVariantCache.getUnchecked(
                                 new AbstractMap.SimpleEntry<>(
                                                t.getTmTextUnitName() + pluralSuffix,
                                                thirdPartyTextUnitForBatchImport.getAsset().getId())
                                );
                            } else {
                                foundCurrentVariant = tmTextUnitCurrentVariantCache.getUnchecked(
This conversation was marked as resolved by chidinmae
                                        new AbstractMap.SimpleEntry<>(
                                                t.getTmTextUnitName(),
                                                thirdPartyTextUnitForBatchImport.getAsset().getId())
                                );
                            }
``` 

is duplicating code, it should be like

```
String textUnitName = t.getTmTextUnitName();

if (pluralFile) {
     textUnitName += pluralSuffix;
}

 foundCurrentVariant = tmTextUnitCurrentVariantCache.getUnchecked(
                                 new AbstractMap.SimpleEntry<>(
                                                textUnitName,
                                                thirdPartyTextUnitForBatchImport.getAsset().getId())
                                );
```

Reply:
got it thanks